lib/hsa/mcwamp_hsa.cpp

```diff
@@ -1590,6 +1590,33 @@ class HSAQueue final : public KalmarQueue
     };
 
 
+    void drain() {
+      while (1) {
+        std::shared_ptr<HSAOp> r;
+        {
+          bool found = false;
+          std::lock_guard<std::recursive_mutex> lg(qmutex);
+          for (int i = 0; i < asyncOps.size(); i++) {
+            if (asyncOps[i] != nullptr) {
+              // wait on valid futures only
+              std::shared_future<void>* future = asyncOps[i]->getFuture();
+              if (future && future->valid()) {
+                r = std::move(asyncOps[i]);
```

Comment:
Do we need to wait on this op to complete before we remove it?  Do we need a future->wait() here ?

Reply:
since @scchan put a `while(1)` loop here I believe his intention is to let those ops complete naturally without explicitly `wait()` on them?

Reply:
the `r.reset()` will destroy the HSAOp and will wait if it is not yet completed.  It occurs outside the lock_guard of the queue to prevent a deadlock.  It keeps going through all the ops until no valid op is found and at that point, it will shrink the asyncOps vector and return.

Reply:
Is there any chance that another Op would be holding a shared_ptr to this Op such that the reset only decrements the ref count and does not destroy/wait on it?